From 5f27ac84ad0db2eb1cdb587401fedd4a3b041527 Mon Sep 17 00:00:00 2001
From: Rami Potinkara <rami.potinkara@qt.io>
Date: Tue, 5 Aug 2025 14:58:37 +0300
Subject: [PATCH] Android: 16KB page size support for building Qt for Android

This patch introduces android_16kb_pages feature.

Feature is needed for Qt versions supporting Android 15 or newer
which use NDK 25 or later.

Feature adds linker options to build shared libraries (.so)
with 16KB page size instead of 4KB page size. The 16KB page size
is compatible with 4KB page size.

Adding the linker flag is applicable only for 64-bit target
architectures: x86_64 and arm64-v8a when using NDK 25, 26 or 27.
NDK 28 and later will have the feature built-in so the linker flags
and note are not needed to be used with those versions.

It is applied to both Qt build and Qt Application builds.

Task-number: QTBUG-131514
Task-number: QTBUG-134093
Pick-to: 6.8 6.5
Change-Id: I8af9c3b392a661995c2eb46b568a40b94ee62aaa
Reviewed-by: Alexandru Croitor <alexandru.croitor@qt.io>
Reviewed-by: Assam Boudjelthia <assam.boudjelthia@qt.io>
(cherry picked from commit d7e0a033c0daa4593462d60def48a897a3d66c6b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit 9da26f935ecb66f874f3e7e92688aeeaefdc7bee)
---
 cmake/QtPlatformTargetHelpers.cmake |  3 +++
 configure.cmake                     | 14 ++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/cmake/QtPlatformTargetHelpers.cmake b/cmake/QtPlatformTargetHelpers.cmake
index 981a64627670..4ea305b8b282 100644
--- a/cmake/QtPlatformTargetHelpers.cmake
+++ b/cmake/QtPlatformTargetHelpers.cmake
@@ -31,6 +31,9 @@ function(qt_internal_setup_public_platform_target)
     # in order to satisfy linker dependencies. Both of these libraries are part of
     # the NDK.
     if (ANDROID)
+        if(QT_FEATURE_android_16kb_pages AND (CMAKE_ANDROID_NDK_VERSION VERSION_LESS "28.0.0"))
+            target_link_options(Platform INTERFACE "-Wl,-z,max-page-size=16384")
+        endif()
         target_link_libraries(Platform INTERFACE log)
     endif()
 
diff --git a/configure.cmake b/configure.cmake
index 9772c2affbe3..f5ce0522f804 100644
--- a/configure.cmake
+++ b/configure.cmake
@@ -1287,6 +1287,13 @@ qt_feature("coverage"
     CONDITION QT_FEATURE_coverage_gcov
 )
 
+qt_feature("android_16kb_pages"
+    LABEL "Using 16KB page sizes in Android"
+    CONDITION ANDROID AND (((CMAKE_ANDROID_NDK_VERSION VERSION_GREATER_EQUAL "25.0.0"))
+                          AND ((CMAKE_ANDROID_ARCH_ABI STREQUAL "arm64-v8a") OR
+                              (CMAKE_ANDROID_ARCH_ABI STREQUAL "x86_64")))
+)
+
 qt_configure_add_summary_build_type_and_config()
 qt_configure_add_summary_section(NAME "Build options")
 qt_configure_add_summary_build_mode(Mode)
@@ -1434,6 +1441,8 @@ qt_configure_add_summary_entry(ARGS "opensslv30")
 qt_configure_add_summary_entry(ARGS "system-zlib")
 qt_configure_add_summary_entry(ARGS "zstd")
 qt_configure_add_summary_entry(ARGS "thread")
+qt_configure_add_summary_entry(ARGS "android_16kb_pages")
+
 qt_configure_end_summary_section() # end of "Support enabled for" section
 qt_configure_add_report_entry(
     TYPE NOTE
@@ -1526,6 +1535,11 @@ qt_configure_add_report_entry(
     MESSAGE "You cannot force both system and bundled libraries."
     CONDITION QT_FEATURE_force_bundled_libs AND QT_FEATURE_force_system_libs
 )
+qt_configure_add_report_entry(
+    TYPE NOTE
+    MESSAGE "Building Qt for Android with 16KB page sizes."
+    CONDITION QT_FEATURE_android_16kb_pages AND (CMAKE_ANDROID_NDK_VERSION VERSION_LESS "28.0.0")
+)
 if(WASM)
     qt_extra_definition("QT_EMCC_VERSION" "\"${EMCC_VERSION}\"" PUBLIC)
 endif()
